### PR TITLE
[Platform] Provide a better error message when the developer certificate can't be used

### PR DESCRIPTION
**Pull request by javiercn**
Partially addresses https://github.com/aspnet/AspNetCore/issues/15118

Long story short, Mac OS will introduce notarization for apps, which changes the identity of `dotnet.exe` with regards to the OS. 

That makes the keys for certs in the cert store (backed by keychain) not accessible to the app if the certificate was created with a `dotnet.exe` that was not notarized.

Users upgrading to 3.1 in the future will encounter this on Mac OS and they won't get any feedback about what's going on other than a failed connection screen in the browser).

To help users in this scenario we are introducing an additional error message with instructions on what to do when we detect the developer certificate is being used for HTTPS, so that at least there is a proper error message.

We are still investigating all the ramifications of this issue, hence the "partially addresses it"

### CODE REVIEW

File: src/Shared/CertificateGeneration/CertificateManager.cs
Comment (javiercn) [inline src/Shared/CertificateGeneration/CertificateManager.cs]: @GrabYourPitchforks I believe this is good enough, isn't it?

File: src/Servers/Kestrel/Core/src/Middleware/HttpsConnectionMiddleware.cs
Comment (Tratcher) [reply src/Servers/Kestrel/Core/src/Middleware/HttpsConnectionMiddleware.cs]: There are likely other ways to cause an AuthenticationException besides a bad dev cert, so we do want to constrain it a bit.

File: src/Servers/Kestrel/Core/src/Middleware/HttpsConnectionMiddleware.cs
Comment (Tratcher) [inline src/Servers/Kestrel/Core/src/Middleware/HttpsConnectionMiddleware.cs]: Add the exception to this log and put the other log into an else clause? No point in logging twice for the same event.

File: src/Servers/Kestrel/Core/src/Middleware/HttpsConnectionMiddleware.cs
Comment (javiercn) [reply src/Servers/Kestrel/Core/src/Middleware/HttpsConnectionMiddleware.cs]: I started in that point, the problem is that I felt we would be spamming people for example when there wasn't a suitable cipher suite, and this guidance would be useless in that case, and I think that's a common enough problem to avoid placing blame on the cert without issue.

What I'm trying to say is that I would be hesitant to provide blanked guidance on blaming the dev-cert for all AuthenticationExceptions as I think we would get misleading bugs in the form, my dev-cert doesn't work and I'm running these two things and doesn't fix it.

Does it make sense?

**Comment by javiercn** [comment]
@Pilchie FYI

File: src/Shared/CertificateGeneration/CertificateManager.cs
Comment (Tratcher) [inline src/Shared/CertificateGeneration/CertificateManager.cs]: Why go through all this work when the candidate was already passed in?

File: src/Servers/Kestrel/Core/src/Middleware/HttpsConnectionMiddleware.cs
Comment (javiercn) [reply src/Servers/Kestrel/Core/src/Middleware/HttpsConnectionMiddleware.cs]: Fair enough

File: src/Shared/CertificateGeneration/CertificateManager.cs
Comment (javiercn) [reply src/Shared/CertificateGeneration/CertificateManager.cs]: @Tratcher because the candidate cert has been "fiddled" with and I haven't been able to access the private key for it. (You can manipulate the instance in memory and have it "detached" from the key).

File: src/Servers/Kestrel/Core/src/CoreStrings.resx
Comment (javiercn) [inline src/Servers/Kestrel/Core/src/CoreStrings.resx]: @danroth27 Are you ok with this message (grammar and all)

File: src/Shared/CertificateGeneration/CertificateManager.cs
Comment (Tratcher) [reply src/Shared/CertificateGeneration/CertificateManager.cs]: How did that happen? Kestrel called the same CertificateManager APIs to get the cert.

Comment (javiercn) [reply src/Shared/CertificateGeneration/CertificateManager.cs]: @Tratcher I'm not 100% sure, it can be that we do something slightly different or that I'm missing a "better pattern" for obtaining the key. This is the best that I've been able to do "empirically". 

Given that this runs on an exception code path and is for a development feature, I'm fine with the current code.

**Comment by javiercn** [comment]
🆙 📅

File: src/Servers/Kestrel/Core/src/Middleware/HttpsConnectionMiddleware.cs
Comment (anurse) [reply src/Servers/Kestrel/Core/src/Middleware/HttpsConnectionMiddleware.cs]: That makes sense! Thanks!

File: src/Shared/CertificateGeneration/CertificateManager.cs
Comment (javiercn) [reply src/Shared/CertificateGeneration/CertificateManager.cs]: I've figured this was a test artifact. The cert is being loaded in a way that makes the private key inaccessible. (I didn't change that, as it benefits my test case).

I've changed the cert in the affected test case for another cert that doesn't pretend to be the aspnet core dev cert, and used that "bad" original cert to add a test case for the error.

Comment (anurse) [reply src/Shared/CertificateGeneration/CertificateManager.cs]: > because the candidate cert has been "fiddled" with and I haven't been able to access the private key for it.

Isn't that exactly what the bug is though? The private key isn't available because of the protections in macOS Catalina? Maybe I'm misunderstanding it.

I do agree that this is not a performance-critical path.

File: src/Servers/Kestrel/Core/src/CoreStrings.resx
Comment (anurse) [inline src/Servers/Kestrel/Core/src/CoreStrings.resx]:
```suggestion
    <value>The ASP.NET Core developer certificate is in an invalid state. To fix this issue, run the following commands 'dotnet dev-certs https --clean' and 'dotnet dev-certs https' to remove all existing ASP.NET Core development certificates and create a new untrusted developer certificate. On macOS and Windows, use 'dotnet dev-certs https --trust' to trust the new certificate.</value>
```

Comment (javiercn) [reply src/Servers/Kestrel/Core/src/CoreStrings.resx]: Thanks!

File: src/Shared/CertificateGeneration/CertificateManager.cs
Comment (GrabYourPitchforks) [reply src/Shared/CertificateGeneration/CertificateManager.cs]: No issues from me. You'll want somebody more familiar with Mac to verify that this doesn't pop UI (e.g., to provide a PIN) on the developer's machine.

**Review by Tratcher — APPROVED** [review]

File: src/Shared/CertificateGeneration/CertificateManager.cs
Comment (javiercn) [reply src/Shared/CertificateGeneration/CertificateManager.cs]: It doesn't pop a UI with the keys we create, but that was a good suggestion.

Thanks for looking into this!

**Comment by javiercn** [comment]
@Pilchie this is just waiting for your approval.

File: src/Shared/CertificateGeneration/CertificateManager.cs
Comment (Pilchie) [inline src/Shared/CertificateGeneration/CertificateManager.cs]: Ugh, public static writable properties?

File: src/Tools/dotnet-dev-certs/src/Program.cs
Comment (Pilchie) [inline src/Tools/dotnet-dev-certs/src/Program.cs]: could be removed.

File: src/Tools/FirstRunCertGenerator/test/CertificateManagerTests.cs
Comment (Pilchie) [inline src/Tools/FirstRunCertGenerator/test/CertificateManagerTests.cs]: _manager could be removed?

**Review by Pilchie — APPROVED** [review]
Doesn't have to be part of this change, but it seems like we should make the constructor of `CertificateManager` private or protected if it's meant to be a singleton.

**Comment by Pilchie** [comment]
Approved for 3.1.0-Preview3.  Thanks @javiercn

File: src/Shared/CertificateGeneration/CertificateManager.cs
Comment (javiercn) [reply src/Shared/CertificateGeneration/CertificateManager.cs]: Yeah, it's a bit ugly, I didn't want to make a big change out of this. 

It's not public code as the class is internal and it is "Shared source". It has become a bit of a mess, over time, but it hasn't reach that point/level where its painful enough to deserve cleanup.

But certainly we can look at cleaning this up as part of 5.0 if we decide to tackle additional things with this tool (like better tracing or adding telemetry to it, which I think would be very useful in the long run (none of our tools currently has))

**Pull request merged**
